### PR TITLE
enable caching via service worker

### LINKED ISSUE #450: AppCache vs. Service Workers
It seems AppCache is getting disliked and Service Workers should be used instead.
https://developers.google.com/web/fundamentals/getting-started/primers/service-workers

### PR DESCRIPTION
**Pull request by peuter**
refs #532 
closes #450

### CODE REVIEW

File: source/ServiceWorker.js [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+var NO_CACHE_TEST = /.+\.php$/i;
+var CONFIG_TEST = /.+visu_config.*\.xml.*/i;
+var config = {};
+var updateQueue = [];
+var queueTid = null;
+
+self.addEventListener('message', function(event) {
+  var data = event.data;
+
+  console.log(data);
```
Comment (ChristianMayer) [inline source/ServiceWorker.js]: On purpose or a debugging-left over that should be removed?

File: source/ServiceWorker.js [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+  if (data.command === "configure") {
+    config = data.message;
+  }
+});
+
+self.addEventListener('install', function(event) {
+  // take over right now
+  console.log("install");
```
Comment (ChristianMayer) [inline source/ServiceWorker.js]: On purpose or a debugging-left over that should be removed?

File: source/ServiceWorker.js [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+});
+
+// delete old caches after activation
+self.addEventListener('activate', function(event) {
+  console.log("activate");
```
Comment (ChristianMayer) [inline source/ServiceWorker.js]: On purpose or a debugging-left over that should be removed?

File: source/ServiceWorker.js [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    return;
+  }
+
+  if (!NO_CACHE_TEST.test(ev.request.url) &&
+      (!ev.request.headers.pragma || ev.request.headers.pragma !== "no-cache") &&
+      (!ev.request.headers['Cache-Control'] || ev.request.headers['Cache-Control'] !== "no-cache")
+  ) {
+
+    ev.respondWith(fromCache(ev.request).then(function(response){
+      // console.log(ev.request.url+" from cache");
```
Comment (ChristianMayer) [inline source/ServiceWorker.js]: missing clean up

File: source/ServiceWorker.js [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+        if (queueTid) {
+          clearTimeout(queueTid);
+        }
+        queueTid = setTimeout(processQueue, 1000);
+      }
+
+      return response;
+    }).catch(function () {
+      // not cached -> do now
+      // console.log("caching " + ev.request.url);
```
Comment (ChristianMayer) [inline source/ServiceWorker.js]: missing clean up

**Comment by peuter** [comment]
I would recommend to not include this PR into the 0.11.0 release. I think we need to gather some experience with the ServiceWorker and it can be used in the best way. As there are so many changes in 0.11.0 I don't think we need another potential source of errors that is included shortly before the release.

**Comment by ChristianMayer** [comment]
OK -> WIP

**Comment by ChristianMayer** [comment]
Is it still WIP or is it ready to be merged?
Anyway there's currently a conflict

**Comment by peuter** [comment]
conflict solved

**Pull request closed without merging**